src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -0,0 +1,462 @@
+                std::to_string(args.size()) + " given)", loc);
+        }
+        ASR::ttype_t *type = ASRUtils::expr_type(args[0]);
+        if (!ASRUtils::is_integer(*type) && !ASRUtils::is_real(*type)
+         && !ASRUtils::is_logical(*type) && !ASRUtils::is_complex(*type)) {
+            err("Argument of the abs function must be Integer, Real, Logical or Complex",
+                args[0]->base.loc);
+        }
+
+        if (is_logical(*type)) {
+            type = TYPE(ASR::make_Integer_t(al, type->base.loc, 4, nullptr, 0));
+        } else if (is_complex(*type)) {
+            type = TYPE(ASR::make_Real_t(al, type->base.loc,
+                ASRUtils::extract_kind_from_ttype_t(type), nullptr, 0));
+        }
+
+        return UnaryIntrinsicFunction::create_UnaryFunction(al, loc, args,
+                eval_Abs, static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Abs),
+                0, type);
+    }
+
+    static inline ASR::expr_t* instantiate_Abs (Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& /*arg_types*/,
+            Vec<ASR::call_arg_t>& new_args, ASR::expr_t* compile_time_value) {
+        ASR::ttype_t* return_type;
+        ASR::symbol_t* original_name;
+        ASR::symbol_t *es = UnaryIntrinsicFunction::import_intrinsic_function(al,
+            loc, scope, "abs", original_name, new_args, return_type);
+// std::cout << ASRUtils::type_to_str_python(return_type) << "\n";
+        LCOMPILERS_ASSERT(es) // TODO
+        return ASRUtils::EXPR(ASR::make_FunctionCall_t(al, loc, es, original_name,
+            new_args.p, new_args.n, return_type, compile_time_value, nullptr));
+    }
+} // namespace Abs
+
+namespace IntrinsicFunctionRegistry {
+
+    static const std::map<int64_t, impl_function>& intrinsic_function_by_id_db = {
+        // LFortran
+        {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::LogGamma),
+            &LogGamma::instantiate_LogGamma},
+        {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Sin),
+            &Sin::instantiate_Sin},
+        {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Cos),
+            &Cos::instantiate_Cos},
+
+        // LPython
+        {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Abs),
+            &Abs::instantiate_Abs}
+    };
+
+    static const std::map<std::string,
+        std::pair<create_intrinsic_function,
+                    eval_intrinsic_function>>& intrinsic_function_by_name_db = {
+                // LFortran
+                {"log_gamma", {&LogGamma::create_LogGamma, &LogGamma::eval_log_gamma}},
+                {"sin", {&Sin::create_Sin, &Sin::eval_Sin}},
+                {"cos", {&Cos::create_Cos, &Cos::eval_Cos}},
+
+                // LPython
+                {"abs", {&Abs::create_Abs, &Abs::eval_Abs}}
+
+    };
+
+    static inline bool is_intrinsic_function(const std::string& name) {
+        return intrinsic_function_by_name_db.find(name) != intrinsic_function_by_name_db.end();
```

Comment:
I would put a check for "sin" here, and ignore everything else for now, and use existing LPython mechanism for those. Let's get "sin" working in this PR, merge it. Then we can add more functions, one by one.